src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -107,35 +223,65 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
         // Calculate cost due to {data point, target neighbors, impostors}
         // triplets.
         double eval = 0;
-        if ((iteration - 1 % range == 0) && (iteration - 1 != 0))
+
+        // Bounds for eval.
+        if (transformationOld.n_elem != 0 && evalOld(l, j, i) < -1)
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 distance(l, i);
+          // Update cache max impostor norm.
+          maxImpNorm(l, i) = std::max(maxImpNorm(l, i), norm(impostors(l, i)));
+
+          eval = evalOld(l, j, i) + transformationDiff *
+              (norm(targetNeighbors(j, i)) + maxImpNorm(l, i) +
+              2 * norm(i));
```

Comment:
Not sure, I get ``norm(i)`` here, it's always ``i`` if ``i >= 0``. Did I miss something? Perhaps the same applies to ``targetNeighbors``.

Reply:
It looks like you got `norm(i)` confused with `arma::norm(i)`(Though I guess arma::norm expects a vector or matrix to passed not a scalar) :)
Actually it is a vector which stores norm of each data point.
`norm(i) = arma::norm(dataset.col(i));`

Reply:
Ahh, this makes sense, thanks for the clarification.